### PR TITLE
Search_for_wildcards function updated to add @DATETO@ functionality

### LINKED ISSUE #508: Search within date range
Hi, 

Currently, one can search for multiple sessions using the @TO@ by looping through session numbers. 

A cool feature would it be if I can specify a date-range, and then download/upload folders with a date within that range. For example, I could specify 20250306@TO@20250309, and get all folders from below except the last one. 

Image

Ideally, this does not require me to have folders for every date range. E.g. I could specify all dates in March, and it would return all the folders in that month even if I did not do recordings every day.

Thanks!

Best
Akseli

### LINKED ISSUE #605: Support `@DATE@, `@TIME@` and `@DATETIME@` to insert values after `sub-` and `ses-` values
Further to #510 (also discussed [here](https://github.com/neuroinformatics-unit/NeuroBlueprint/issues/74)) it would be nice to support outright the `sub-<date>` (or `ses-`, `<time>`, `<datetime>`) format. The two things to do are filter them as part of #510 and patch `@DATE@`, `@TIME@` and `@DATETIME@` to insert correctly in after `sub-` and `ses-` keys.

Otherwise, these should work outright. The only other issues is that for `@DATETIME@` you would need to set `allow_letters_in_sub_ses_values` which could be confusing, but I think we can handle that if it is raised as an issue.

### PR DESCRIPTION
**Pull request by Diya910**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/neuroinformatics-unit/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [ yes] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
This PR introduces a new @DATETO@ wildcard that enables users to search for folders based on a date range embedded in their names. This feature is especially useful when users want to transfer data recorded within a specific date range, without needing to create folders for every date in that range.

**What does this PR do?**
Implements @DATETO@ pattern recognition inside search_for_wildcards.

Uses get_values_from_bids_formatted_name to extract date-YYYYMMDD from folder names.

Filters the folders based on whether the date falls within the provided range.

## References
#508 

## How has this PR been tested?
Created automated tests (test_date_search_range) using a simulated folder structure with date-YYYYMMDD format.

Verified that only folders within the specified date range are returned.

Confirmed that existing wildcard functionality remains unaffected.

## Is this a breaking change?
No, this feature is additive and does not alter existing behavior.

## Does this PR require an update to the documentation?
Yes. The documentation should be updated to mention the new @DATETO@ wildcard and its usage.

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [ yes] The code has been tested locally
- [ yes] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

There are two minor mypy errors I couldn't fully resolve:
A type conflict involving the dummy Configs class used in tests — guidance from maintainers would help finalize this.
A type mismatch originating from an existing code path — this appears unrelated to the new functionality added.

### CODE REVIEW

**Comment by Diya910** [comment]
@adamltyson @JoeZiminski 
Is there any update on the pull request. Your feedback will be really helpful.

**Comment by JoeZiminski** [comment]
Hi @Diya910 so sorry for the delay in response! thanks a lot for this PR and the extensive tests. I'm still not back full time but will definitely have time to review this within the next two weeks. Thanks for your patience

File: datashuttle/utils/folders.py [hunk 1]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
...
     for name in all_names:
-        if canonical_tags.tags("*") in name:
-            name = name.replace(canonical_tags.tags("*"), "*")
-
-            matching_names: List[str]
+        if canonical_tags.tags("*") in name or "@DATETO@" in name:
+            search_str = name.replace(canonical_tags.tags("*"), "*")
+            # If a date-range tag is present, extract dates and update the search string.
+            if "@DATETO@" in name:
+                m = re.search(r"(\d{8})@DATETO@(\d{8})", name)
```
Comment (JoeZiminski) [inline datashuttle/utils/folders.py:411]: This is nice, we have some validation code for ISO formats [here](https://github.com/neuroinformatics-unit/datashuttle/blob/b82cf9ee62f0a69b137b2954f31245789050a45c/datashuttle/utils/validation.py#L428C5-L428C28) however this function works on a list of names, and returns in a slightly strange format (which makes sense in the context of the validation functions).

I think to centralise this we can do a few refactorings:

Move this code block:
```
    formats = {
        "datetime": "%Y%m%dT%H%M%S",
        "time": "%H%M%S",
        "date": "%Y%m%d",
    }
```   

to `configs/canonical_tags` and wrap it in a function like `get_datetime_format` based on the format.

Then this code in `datetimes_in_iso_format` can be factored into a separate function `valdiate_datetime`

```
        strfmt = formats[key]

        try:
            datetime.strptime(format_to_check, strfmt)
            error_message = []
        except ValueError:
            error_message = [get_datetime_error(key, name, strfmt, path_)]
 ```
except it can just return True / False and we can leave the error message stuff to `datetimes_in_iso_format`.

Now, we can call this new function from here. I think it is also worth having a quick function to get the expected number of values (8 for date) for use above, instead of hard-coding.  This could be like:

```
def get_expected_num_datetime_values(format):
    format_str = get_datetime_format(format)
    today = datetime.now()
    return len(today.strftime(format_str)
```

We can then pass this "date", "time" or "datetime" to get the values.

This section would then look something like:

```
# somewhere need to check that @DATETO@, @TIMETO@ and @DATETIMETO@ are used exclusively
format = tag  = None
if tags.tag("DATETO") in search_str:
    format = "date"
    tag = tags.tag("DATETO") 
elif  tags.tag("TIMETO" in search_str:
    format = "time"
    tag = tags.tag("TIMETO") 
elif tags.tag("DATETIMETO") in search_str:
    format = "datetime"
    tag = tags.tag("DATETIMETO") 
    
expected_values =  get_expected_num_datetime_values(format)
full_tag_regex = fr"(\d{{{num_values}}}){re.escape(tag)}(\d{{{num_values}}})"
match = re.search(full_tag_regex , search_str)

if not match:
...raise (use utils. raise_error and raise a NeuroBlueprint error)

start_str, end_str = match.groups()

start_timepoint =datetime.strptime(start_str, get_datetime_format(format)")
end_timepoint =datetime.strptime(end, get_datetime_format(format)")
    
if not validate_datetime(start_timepoint, format):
... raise error

<same for end_timepoint>

search_str = re.sub(full_tag_regex , f{format}-*", search_str)
```

I think all of this could be isolated to a new function such that in this function we just have something like:
```
format = tag  = None
if tags.tag("DATETO") in search_str:
    format = "date"
    tag = tags.tag("DATETO") 
elif  tags.tag("TIMETO" in search_str:
    format = "time"
    tag = tags.tag("TIMETO") 
elif tags.tag("DATETIMETO") in search_str:
    format = "datetime"
    tag = tags.tag("DATETIMETO") 
    
search_str = format_and_validate_datetime_search_str(search_str, format, tag)
 ```

File: datashuttle/utils/folders.py [hunk 2]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
...
                 )[0]
             else:
-                matching_names = search_sub_or_ses_level(  # type: ignore
-                    cfg, base_folder, local_or_central, search_str=name
+                matching_names = search_sub_or_ses_level(
+                    cfg, base_folder, local_or_central, search_str=search_str
                 )[0]
-
+            # If a date-range tag was provided, further filter the results.
+            if "@DATETO@" in name:
```
Comment (JoeZiminski) [inline datashuttle/utils/folders.py:438]: This is nice, and at this point we know we have validated dates. Ideally the validation should happen immediately before the point of use but in this case, there is no point wasting time searching if the dates are not valid, so it makes sense to do it before. But it is worth leaving a comment to indicate we know the dates are valid at this stage.

File: datashuttle/utils/folders.py [hunk 3]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
...
+                filtered_names: List[str] = []
+                for candidate in matching_names:
```
Comment (JoeZiminski) [inline datashuttle/utils/folders.py:440]: This is nice, to generalise it a bit more we can have:

```
if format is not None:
    assert tag is not None, "format and tag should be set together"
    
get_values_from_bids_formatted_name can use `format` and in the strptime call use the new `get_datetime_format ` function

```

File: datashuttle/utils/folders.py [hunk 4]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
...
+                    candidate_basename = (
+                        candidate
+                        if isinstance(candidate, str)
+                        else candidate.name
+                    )
+                    values_list = get_values_from_bids_formatted_name(
+                        [candidate_basename], "date"
+                    )
+                    if not values_list:
```
Comment (JoeZiminski) [inline datashuttle/utils/folders.py:449]: I think we can assume this list is not empty because `date-*` was used to search the names already (?)

File: datashuttle/utils/folders.py [hunk 5]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
...
+                        continue
+                    candidate_date_str = values_list[0]
+                    try:
+                        candidate_date = datetime.strptime(
+                            candidate_date_str, "%Y%m%d"
+                        )
+                    except ValueError:
+                        continue
+                    if start_date <= candidate_date <= end_date:
+                        filtered_names.append(candidate)
```
Comment (JoeZiminski) [inline datashuttle/utils/folders.py:459]: I think this entire block could be isolated in a new function ```filter list of names by datetime``` just for readability (and it might also be useful in future)

**Review by JoeZiminski — CHANGES_REQUESTED** [review]
Hey @Diya910 thanks a lot for this, its a really nice implementation and is exactly what we need to do in this case. I have left a few comments on refactoring, this is because the introduced functionality can be aligned with some existing code to reduce duplication across the codebase. This requires some massaging of existing datashuttle code to make it a little more general so it can be called here. The suggestions also extend the implementation to handle the TIMETO and DATETIMETO case. For now I have not reviewed the tests as they might need changing after the refactor, but in general they look good and the attention to detail on testing is much appreciated.

Let me know if anything is not clear and if you have any questions or alternative ways to tackle this. Refactorings like those suggested can be a little fiddly. The linting / type checking will be useful when performing such refactorings. Of course, I'm happy to help wherever it would be useful. Thanks again for this contribution!

Just a reminder to myself, we will also need to add documentation for this new functionality.

**Comment by JoeZiminski** [comment]
Hey @Diya910 great! No rush BTW I was just checking in, please prioritise exams / other stuff / taking some time to recuperate after exams. I was thinking it might be nice to merge over the next few weeks (rather than next few days), thanks!

File: datashuttle/utils/folders.py [hunk 6]
```diff
@@ -401,27 +404,65 @@ def search_for_wildcards(
     """
     new_all_names: List[str] = []
     for name in all_names:
-        if canonical_tags.tags("*") in name:
-            name = name.replace(canonical_tags.tags("*"), "*")
-
-            matching_names: List[str]
+        if canonical_tags.tags("*") in name or "@DATETO@" in name:
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py]: Reorganized the code to handle wildcards and datetime tags separately

File: datashuttle/utils/folders.py (hunk 1 above)
Comment (Diya910) [reply datashuttle/utils/folders.py]: Moved datetime formats to canonical_tags.py and created get_datetime_format function for centralized access

File: datashuttle/utils/folders.py (hunk 5 above)
Comment (Diya910) [reply datashuttle/utils/folders.py]: Created filter_names_by_datetime_range() function

**Review by JoeZiminski — CHANGES_REQUESTED** [review]
Hey @Diya910 thanks for this, this is really great stuff. The code is very clean, this is going to make a great feature. I have left a few comments on the code, they just suggest some minor refactoring's to reduce code duplication where possible. For critical code, it makes sense to define the key parts only in once place, just in case they are changed later but the editor forgets to check for all places they are defined.

The tests are great for ensuring the features works well, I have suggested a refactoring here to use our existing testing machinery which I think should reduce some boilerplate, let me know if you have any questions about this. The tests will should probably test all three cases, dateto, timeto and datetimteto, happy to help with this.

I just pushed some fixes to the pre-commit on the CI which was failing, just some minor typing issues (see [here](https://datashuttle.neuroinformatics.dev/pages/community/contributing.html#formatting-and-pre-commit-hooks) for some detail on the pre-commit hooks). This should move on to the full test suite now.

Thanks again Diya this is nearly done! I just remembered we will also need to document this change, the contributing guide for this is [here](https://datashuttle.neuroinformatics.dev/pages/community/contributing.html#contributing-documentation). It would make sense to add the new tags to [this section](https://datashuttle.neuroinformatics.dev/pages/user_guides/transfer-data.html#custom-transfer-keywords). Happy to do this because the documentation can be a bit fiddly, but if you are interested in this please feel free to go ahead, let me know if you have any questions!

File: datashuttle/utils/validation.py [hunk 7]
```diff
@@ -432,18 +436,13 @@ def datetime_are_iso_format(
...
     """
-    formats = {
-        "datetime": "%Y%m%dT%H%M%S",
-        "time": "%H%M%S",
-        "date": "%Y%m%d",
-    }
-
-    key = next((key for key in formats if key in name), None)
+    key = next(
+        (key for key in ["datetime", "time", "date"] if key in name), None
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/validation.py:440]: datetime_keys = list(canonical_tags.get_datetime_format())
key = next((key for key in datetime_keys if key in name), None)
 Like this??

File: datashuttle/utils/folders.py [hunk 8]
```diff
@@ -359,7 +362,53 @@ def process_glob_to_find_datatype_folders(
...
+            candidate if isinstance(candidate, str) else candidate.name
+        )
+        value = get_values_from_bids_formatted_name(
+            [candidate_basename], format_type
+        )[0]
+        try:
+            candidate_timepoint = datetime.strptime(
+                value, canonical_tags.get_datetime_format(format_type)
+            )
+            if start_timepoint <= candidate_timepoint <= end_timepoint:
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:404]: Should I remove the whole try/ except block or just move the if statement outside it? Also I would really like to know the reason behind doing this.

File: datashuttle/utils/folders.py [hunk 9]
```diff
@@ -397,31 +446,82 @@ def search_for_wildcards(
...
+        # Initialize search string
+        search_str = name
+
+        # Handle wildcard replacement first if present
+        if canonical_tags.tags("*") in name:
+            search_str = search_str.replace(canonical_tags.tags("*"), "*")
+
+        # Handle datetime ranges
+        format_type = tag = None
+        if canonical_tags.tags("DATETO") in search_str:
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:470]: This was something new for me. I'll do it.

File: datashuttle/utils/validation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoeZiminski) [reply datashuttle/utils/validation.py:440]: Oh I see apologies for the confusion, the return value of ```get_datetime_format``` is slightly different to that I thought.

What do you think of changing the function to take no arguments and return the entire dictionary, and then it can be indexed (instead of called). For example `format = get_datetime_format(format_type)` becomes `format = get_datatime_formats()[format_type]`.

Then above, the signature can be:
```
datetime_keys = list(canonical_tags.get_datetime_formats().keys())
key = next((key for key in datetime_keys if key in name), None)
```

This signature is now different to the `tags()` function, but it is probably a better design because it is more flexible. Now the datetime names are canonically defined in a central place, and we can grab them all or index them as we like. Later on `tags()` can be changed to follow the same design. Sorry this would be a bit of a pain to change all the calls in your code though. What do you think?

File: datashuttle/utils/folders.py [hunk 10]
```diff
@@ -397,31 +446,82 @@ def search_for_wildcards(
...
+            matching_names = search_sub_or_ses_level(
+                cfg, base_folder, local_or_central, search_str=search_str
+            )[0]
+
+        # Filter results by datetime range if one was present
+        if format_type is not None and tag is not None:
+            expected_values = validation.get_expected_num_datetime_values(
+                format_type
+            )
+            full_tag_regex = rf"(\d{{{expected_values}}}){re.escape(tag)}(\d{{{expected_values}}})"
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:507]: I couldn't get it clearly. Can you elaborate this?

File: datashuttle/utils/validation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/validation.py:440]: Thanks, that makes a lot of sense! I agree that returning the full dictionary from get_datetime_formats() would make the design cleaner and more flexible — especially for cases like this where we need all the keys. Happy to refactor the calls where needed. Also agree that aligning the design of tags() later on would help keep things consistent across the codebase. Will go ahead with this change!

File: tests/test_date_search_range.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
+import glob
```
[the remark this answers is not in the document]
Comment (JoeZiminski) [reply tests/test_date_search_range.py:1]: Hi @Diya910, I think I accidentally copied from a commit, does [this](https://github.com/neuroinformatics-unit/datashuttle/blob/main/tests/tests_integration/test_filesystem_transfer.py#L343) work?

it is the function `test_wildcard_transfer` in `/tests/tests_integration/test_filesystem_transfer.py`

File: datashuttle/utils/validation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoeZiminski) [reply datashuttle/utils/validation.py:440]: great!

**Comment by Diya910** [comment]
Hey @JoeZiminski, I have done changes required by you. These were a lot of changes I am not able to reply to all of them individually. But I made sure to make changes suggested by you. I have tested the changes on draft test file and they are working fine. I haven't properly done work on test file. It was a lot for me to do in a go. Once you confirm these changes I'll move ahead in refactoring test file. I hope you are fine with it. If I missed any suggestion above just in case, please point out to that I'll make those changes.

**Review by JoeZiminski — CHANGES_REQUESTED** [review]
Hey @Diya910 this is great, definitely good to go bar some very minor suggestions. Most of these are minor github code suggestions so you can directly commit them.

Apologies, one of my suggestions was actually worse than what was already there 😅 around the walrus operator. Sorry for the inconvenience of having to revert this.

After these changes are integrated I will message @Akseli-Ilmanen to test this manually while the other tests are been written. Let me know if you have any questions as you refactor the tests. Thanks again!

File: datashuttle/configs/canonical_tags.py [hunk 12]
```diff
@@ -11,5 +11,43 @@ def tags(tag_name: str) -> str:
...
+        A dictionary containing format strings for datetime, time, and date
+    """
+    return {
+        "datetime": "%Y%m%dT%H%M%S",
+        "time": "%H%M%S",
+        "date": "%Y%m%d",
+    }
+
+
+def get_datetime_format(format_type: str) -> str:
```
Comment (JoeZiminski) [inline datashuttle/configs/canonical_tags.py:37]: here sometimes in the code `get_datetime_format(format)` is used and other places `get_datetime_formats()[format]`. I think we can remove `get_datetime_format` and always use  `get_datetime_formats()[format]`

File: datashuttle/utils/folders.py [hunk 13]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
-    but where @*@-containing names have been replaced with
-    search results.
+    For datetime ranges, the format must be:
+    - date: YYYYMMDD@DATETO@YYYYMMDD (e.g., "20240101@DATETO@20241231")
+    - time: HHMMSS@TIMETO@HHMMSS (e.g., "000000@TIMETO@235959")
+    - datetime: YYYYMMDDTHHMMss@DATETIMETO@YYYYMMDDTHHMMss
 
     Parameters
     ----------
+    cfg : Configs
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:387]: All declarations?

File: datashuttle/utils/folders.py [hunk 14]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
-            matching_names: List[str]
+        # Handle wildcard replacement first if present
+        search_str = name
+        if canonical_tags.tags("*") in name:
+            search_str = search_str.replace(canonical_tags.tags("*"), "*")
+
+        # Handle datetime ranges
+        format_type = None
+        tag = None
+        if (tag := canonical_tags.tags("DATETO")) in search_str:
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:445]: No worries. It was my fault too. I didn't catch it.

File: datashuttle/utils/folders.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:387]: so you want me to write like cfg only correct? Obviously wherever I have done declarations in the code.

File: datashuttle/utils/folders.py [hunk 15]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
-        may or may not include the wildcard flag. If sub (below)
-        is passed, it is assumed these are session names. Otherwise,
-        it is assumed these are subject names.
-
-    sub : optional subject to search for sessions in. If not provided,
-        will search for subjects rather than sessions.
-
+    Returns
+    -------
+    List[str]
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:403]: Okay, got it!!

File: datashuttle/utils/folders.py [hunk 16]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
-    )  # remove duplicate names in case of wildcard overlap
+    Parameters
+    ----------
+    names : List[str]
+        List of names to filter, all containing the datetime pattern
+    format_type : str
+        One of "datetime", "time", or "date"
+    start_timepoint : datetime
+        Start of the datetime range
+    end_timepoint : datetime
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:508]: Okay let me do.

File: datashuttle/utils/folders.py [hunk 17]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
+    return start_timepoint, end_timepoint
+
+
+def format_and_validate_datetime_search_str(search_str: str, format_type: str, tag: str) -> str:
+    """
+    Validate and format a search string containing a datetime range.
+
+    Parameters
+    ----------
+    search_str : str
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:654]: Yes, sure

File: datashuttle/utils/folders.py [hunk 18]
```diff
@@ -355,74 +357,352 @@ def process_glob_to_find_datatype_folders(
...
+        The formatted search string with datetime range replaced
+        e.g. "sub-001_date-*_id-*" or "sub-002_time-*"
+
+    Raises
+    ------
+    NeuroBlueprintError
+        If the datetime format is invalid or the range is malformed
+    """
+    # Extract and validate datetime range
+    strip_start_end_date_from_datetime_tag(search_str, format_type, tag)
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:674]: Yes, I have changed it to validate datetime range format

File: datashuttle/configs/canonical_tags.py (hunk 12 above)
Comment (Diya910) [reply datashuttle/configs/canonical_tags.py:37]: Ahm!! sorry for this inconsistency. Let me fix it

**Comment by JoeZiminski** [comment]
Hey @Diya910 thanks for these changes, all looks great! Docstrings are formatted correctly, I think good to move on to tests now! Let me know if you have any questions about anything.

Currently the `pre-commit` test is failing, you can read some details on our tools [here](https://datashuttle.neuroinformatics.dev/pages/community/contributing.html#formatting-and-pre-commit-hooks). Basically, you can do `pre-commit install` in the environment, then every time you commit it will run all the checks. Let me know if you have any questions about solving the issues, sometimes it is not clear. You can also do `pre-commit run --all-files` to run on the entire codebase (which  you can do now to fix the issues). You can see the issues in the Checks section at the bottom of this page.

At some point I will merge the new main into your branch and fix the conflicts (I think this will be easier if I do it, as I know the changes that are coming in from main branch so will be easier to solve). Let me know a day that you will not be using the branch and I can do it then.

The final thing to do will be to add this feature to the documentation, see a guide [here](https://datashuttle.neuroinformatics.dev/pages/community/contributing.html#contributing-documentation). If you would like to learn the sphinx documentation stack, please give it a go and ask me any questions you have. Otherwise, I'm happy to do it.

**Comment by JoeZiminski** [comment]
Hey @Diya910 sure I have merged main into this branch, there were a lot of fiddly conflicts that would be basically impossible to solve if you had not worked on the main branch changes. That's fixed now, I also made some changes related to typing to ensure the linter passes. Let me know if you have any questions!

**Comment by Diya910** [comment]
Yeah sure, no worries. I will be happy to work.

**Review by JoeZiminski — CHANGES_REQUESTED** [review]
Hey @Diya910 thanks again for these tests, they are excellent and cover not only the core cases but many useful additional cases. Please see for a couple of suggestions, there is one edge case in which the bounds can be adjusted, and the regexp for the datetime needs to be slightly adjusted too along with the corresponding tests. Once these changes are made I think all changes to the code will be done! Then just the documentation, I think we only really need a few lines there, and this will be good to go. Thanks again for this great contribution, this is a really cool feature.

**Comment by Diya910** [comment]
Hey @JoeZiminski, yes you may do the further changes. I was planning to do this weekend. But if you want to take the charge you may proceed. Thankyou

**Comment by Diya910** [comment]
Hey @JoeZiminski , I was trying to work yesterday on the PR but there are merging conflicts with main branch. Can you please resolve them so that I can work further.

**Comment by Akseli-Ilmanen** [comment]
@JoeZiminski, thanks for the explanation on how to manage installation, very helpful!


Would your example work? I thought that the date has to be specified in YYYYMMDD format?
> 
> To use the new feature, you can add a `@DATETO@`, `@TIMETO@` or `@DATETIMETO@` surrounded by the time range you want to restrict to e.g. `sub-@*@_01012025@DATETO02022025`

**Comment by Akseli-Ilmanen** [comment]
Ah oopsie. I fixed the @, but still get these errors: 

When I run:
```
from datashuttle import DataShuttle

project = DataShuttle("AI_data")
project.upload_custom(
    top_level_folder="derivatives",
    sub_names="all_sub",
    ses_names="ses-@*@_20250309@DATETO@20250310",
    datatype="behav"
)
```
I get:

```
The sub names to transfer are: ['sub-01_id-Ivy', 'sub-02_id-Poppy', 'sub-03_id-Freddy']
The ses names to transfer are: []
The ses names to transfer are: []
The ses names to transfer are: []
No files included. None transferred.
```

And if I run with the `ses-000` modification: 

```
from datashuttle import DataShuttle


project = DataShuttle("AI_data")
project.upload_custom(
    top_level_folder="derivatives",
    sub_names="all_sub",
    ses_names="ses-000_20250309@DATETO@20250310",
    datatype="behav"
)
```
I get this error message:

```
The sub names to transfer are: ['sub-01_id-Ivy', 'sub-02_id-Poppy', 'sub-03_id-Freddy']
---------------------------------------------------------------------------
NeuroBlueprintError                       Traceback (most recent call last)
Cell In[11], [line 5](vscode-notebook-cell:?execution_count=11&line=5)
      1 from datashuttle import DataShuttle
      4 project = DataShuttle("AI_data")
----> [5](vscode-notebook-cell:?execution_count=11&line=5) project.upload_custom(
      6     top_level_folder="derivatives",
      7     sub_names="all_sub",
      8     ses_names="ses-000_20250309@DATETO@20250310",
      9     datatype="behav"
     10 )

File ~\Documents\Akseli\Code\datashuttle\datashuttle\utils\decorators.py:66, in check_configs_set.<locals>.wrapper(*args, **kwargs)
     60     log_and_raise_error(
     61         "Must set configs with make_config_file() "
     62         "before using this function.",
     63         ConfigError,
     64     )
     65 else:
---> [66](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/Admin/Documents/Akseli/Code/movformer-gui/~/Documents/Akseli/Code/datashuttle/datashuttle/utils/decorators.py:66)     return func(*args, **kwargs)

File ~\Documents\Akseli\Code\datashuttle\datashuttle\utils\decorators.py:88, in check_is_not_local_project.<locals>.wrapper(*args, **kwargs)
     81     log_and_raise_error(
     82         "This function cannot be used for a local-project. "
     83         "Set connection configurations using `update_config_file` "
...
     80 """
     81 ds_logger.close_log_filehandler()
---> [82](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/Admin/Documents/Akseli/Code/movformer-gui/~/Documents/Akseli/Code/datashuttle/datashuttle/utils/utils.py:82) raise exception(message)

NeuroBlueprintError: SPECIAL_CHAR: The name: ses-000_20250309@DATETO@20250310, contains characters which are not alphanumeric, dash or underscore.
Output is truncated. View as a [scrollable element](command:cellOutput.enableScrolling?b37f3fe7-1f32-449c-9835-1522309136b4) or open in a [text editor](command:workbench.action.openLargeOutput?b37f3fe7-1f32-449c-9835-1522309136b4). Adjust cell output [settings](command:workbench.action.openSettings?%5B%22%40tag%3AnotebookOutputLayout%22%5D)...
```

File: datashuttle/utils/folders.py [hunk 19]
```diff
@@ -453,39 +501,357 @@ def search_for_wildcards(
...
+    ------
+    NeuroBlueprintError
+        If the datetime format is invalid or the range is malformed
+
+    """
+    # Validate the datetime range format
+    strip_start_end_date_from_datetime_tag(search_str, format_type, tag)
+
+    # Replace datetime range with wildcard pattern
+    expected_len = get_expected_datetime_len(format_type)
```
[the remark this answers is not in the document]
Comment (Diya910) [reply datashuttle/utils/folders.py:790]: There will actually be 3 types because date will have 8 digit so "\d{8}". Am I right? Should I change it accordingly?

**Comment by JoeZiminski** [comment]
Hi @Akseli-Ilmanen thanks for this, that' is very useful to know. Great point that it gets very difficult to track session numbers as the experiment run for longer and /or are more complex. The `ses-<datetime>_n-<integer>` format is definitely NeuroBlueprint and should work without issue in datashuttle. 

One problem is the `@DATE@` shortcut won't work, as it tries to add a `date` key like `sub-date-<date>`, but we can patch this to not insert the `date` key in this situation. Alongside a patch in this PR this should provide the full functionality to the `ses-<datetime>` pattern.

I'll make an issue for the above and try and prioritise this over the next 2-3 weeks, will keep you updated! Thanks again for all the valuable feedback.

**Comment by Akseli-Ilmanen** [comment]
This is great, looking forward to it, thanks!

**Comment by JoeZiminski** [comment]
Hey @Akseli-Ilmanen thanks for these suggestions, on this PR it should now be possible to do things like:

```
from datashuttle import DataShuttle

project = DataShuttle("my_project")

project.create_folders("rawdata", "sub-@DATE@", "ses-@DATETIME@", ["behav", "ephys"], allow_letters_in_sub_ses_values=True)

# collect some data

project.upload_custom("rawdata", "all", "ses-20251030T134534@DATETIMETO@20251030T134544", "all")
```

And equivilently this can be done through the TUI. Note that for the transfer to work, every `ses-` must have the same format (e.g. all must be `ses-<a datetime>` or `ses-<a date>` or `ses-<a time>` but not a mix (similar for `sub-`). 

Would be great to hear how this works for you!

You can install from this PR by doing:

```
pip uninstall datashuttle
git clone git@github.com:Diya910/datashuttle.git
cd datashuttle
git checkout date_feature
git pull
pip install -e .
```

**Review by JoeZiminski — APPROVED** [review]
This is good to go, thanks so much @Diya910! This is a great addition to the package
